### ISSUE TITLE
Some questions about the understanding of your codes

### ISSUE THREAD

**Issue by hnyz979** [bryanyzhu/deepOF#4]
Dear Dr.Zhu
  I have read through your paper and codes using Sintel for training last week. In your paper, FlowFields is used as the classical optical flow estimator to provide the proxy ground truth. This proxy groundtruth is used to train the CNN-based optical flow estimators. According to my understanding, it seems that the proxy groundtruth are the **pr1~6** in the codes, for they are used as the **"flows"** in the function "loss_interp_multi(**flows**, inputs, epsilon, alpha_c, alpha_s, lambda_smooth, flow_scale, deltaWeights)", as shown in the codes below:
````
(in function "inception_v3" in sintelWrapFlow.py): 
loss6, _ = loss_interp_multi(pr6, pr6_input, epsilon, alpha_c, alpha_s, lambda_smooth, flow_scale_6, deltaWeights)

(in function "loss_interp_multi" in sintelWrapFlow.py):
def loss_interp_multi(flows, inputs, epsilon, alpha_c, alpha_s, lambda_smooth, flow_scale, deltaWeights):
...
flows = tf.multiply(flows, flow_scale)
flows_flat = tf.reshape(flows, [num_batch, -1, 2])
floor_flows = tf.to_int32(tf.floor(flows_flat))
weights_flows = flows_flat - tf.floor(flows_flat)
````

However, it seem that the **pr6** is computed from end_points["Mixed_7c"], as shown in the codes below:
````
(in function "inception_v3" in sintelWrapFlow.py): 
pr6 = slim.conv2d(end_points["Mixed_7c"], 2*(time_step-1), [3, 3], activation_fn=None, scope='pr6')
````

 However, end_points["Mixed_7c"] is obtained using function "inception_v3_base". It seems that function "inception_v3_base" is not an "non-CNN based classical methods", because convolution and filtering layers are used in this function. Could you help me find where did I mis-understand your code? Thank you very much!

**Comment by hnyz979** [comment]
Some detailed questions in the "loss_interp_multi" function:
(1)In the code (in loss_interp_multi function in sintelWrapFlow.py):
` x = floor_flows[b, :, idx1]    # U, horizontal displacement`
What does the x stand for in this code?  I think it may be the horizontal displacement calculated by the optical flow in one frame. But I do not quite understand why (because the "floor_flows" seems to be pr6~pr1, which are calculated using slim.conv2d function)?

(2)In the codes (in loss_interp_multi function in sintelWrapFlow.py):
```
floor_flows = tf.to_int32(tf.floor(flows_flat))
weights_flows = flows_flat - tf.floor(flows_flat)
...
wa = (1-xw) * (1-yw)
wb = (1-xw) * yw
wc = xw * (1-yw)
wd = xw * yw
```
It seem that your paper did not run into these details of the implementation, so I did not quite understand what these codes actually mean (although I roughly guess that it may be used to reconstruct the I1' from I2 and the optical flow). Could I know where can I find these details?
thanks~

**Comment by bryanyzhu** [comment]
Hi, the code above is used to interpolate the image in local region. As a simple example for linear interpolation, you will need 4 points nearby to reconstruct the point on the grid. The weights_flows are the distance, wa,wb,wc,wd are the weights during interpolation. You can refer to spatial transformer network for more details, and [here](https://github.com/kevinzakka/spatial-transformer-network) is its Tensorflow implementation. Hope it helps.

**Comment by hnyz979** [comment]
Thank you very much, especially the reference of the spatial transformer network. I found in that work that the "transformer.py" is quite the same with your codes. 
There are still two questions that I do not quite understand, as follows:
(1) When grabbing the 4 nearest corner points, spatial transformer network uses:
(lines 198-201 in the "transformer.py" code)

````
x0 = tf.cast(tf.floor(x), 'int32')
 x1 = x0 + 1
 y0 = tf.cast(tf.floor(y), 'int32')
y1 = y0 + 1
````
while in our optical flow case, it changes to: 
(lines 550-553 in your "sintelWrapFlow.py" code)
````
x0 = pos_y + x
x1 = x0 + 1
y0 = pos_x + y
y1 = y0 + 1
````
why should x0 be calculated by pos_y (but not pos_x)?

(2) In lines 545-548 in your "sintelWrapFlow.py" code, why should idx2 be used to get y and yw?

```
x = floor_flows[b, :, idx1]    # U, horizontal displacement
y = floor_flows[b, :, idx2]    # V, vertical displacement
xw = weights_flows[b, :, idx1]
yw = weights_flows[b, :, idx2]
```

It seems to me that the optical flow of the x and y direction should be taken from the same direction. so I am still confused here. Anything wrong with my understanding?


(3) Which codes corresponds to the equation (1) in your paper? It seem that the pr6-pr1 are all the calculated optical flows, and loss6-loss1 are all reconstructions losses (equation 2 in your paper), but I did not see the EPE error is used to calculate the loss in "sintelWrapFlow.py" code.

Thank you very much.

**Comment by bryanyzhu** [comment]
Hi, 
(1) Because the way we build the grid is different. Ours is like [this](https://github.com/bryanyzhu/deepOF/blob/master/sintelWrapFlow.py#L528). So posx actually refers to height, which is flow in the y direction.  It is kind of our variable misnaming, sorry about the confusion. I suggest you follow transformer.py. 
(2) Actually idx1 is 0, idx2 is 1, which corresponds to flow in the x direction and y direction. 
(3) This repo is not completed, so there is no EPE calculation. This repo only supports unsupervised loss. But if you want to compute EPE, it should be very simple. Because EPE is just MSE loss. 

Hope it helps. Thanks.

**Comment by hnyz979** [comment]
Thank you for your time and patience, Dr. Zhu!
